### PR TITLE
MarketTrend: add new alpha panel

### PR DESCRIPTION
**Pull request by leeoniya**
the Market Trend panel builds on top of the Time series panel, and adds special rendering for mapped open, high, low, close, volume fields. (this is a continuation of https://github.com/grafana/grafana/pull/36406)

image

**Highlights**

- candlesticks, both solid and hollow variants
- OHLC bars
- customizeable up/down colors
- candle or bar color determined by either intra-period or inter-period price movement
- volume histogram with matched colors
- volume histogram can be detached into separate panel for flexible dashboarding, panel alignment and additional up/down color customization
- all fields that are not mapped for special rendering have full Time series styling and overrides available, including e.g. `fillBelowTo` for Bollinger bands construction.
- logarithmic scales and other features work the same as they do in the Time series panel

**Naming is Hard™**

Why not "Candlestick", "Financial", "Stock", "Market", "K Line" panel?

1. OHLC Bars that this panel can render are not candlesticks
2. In volume-only mode, there are no candlesticks or bars rendered
3. "Financial" is too broad, there are many of those.
4. Things like crypto coins are not "stocks"
5. "Market" alone is also too broad; i expect a [Market Depth](https://www.google.com/search?q=market+depth+chart&tbm=isch) (or Order Book) vis to follow in the future
6. "K Line" is too obscure for someone outside the industry to know what it is

I also have some concern about the "price*" nomenclature for panel options, since Forex trading and other arbitrage charts have a ratio between prices on the Y axis (unitless).

Inspiration for how a financial dashboard could be assembled in the future is taken from this image:

image

### CODE REVIEW

File: public/app/plugins/panel/market-trend/MarketTrendPanel.tsx [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+  return (
+    <TimeSeries
+      frames={frames}
+      structureRev={data.structureRev}
+      timeRange={timeRange}
+      timeZone={timeZone}
+      width={width}
+      height={height}
+      legend={options.legend}
+      renderers={renderers as any}
```
Comment (ryantxu) [inline public/app/plugins/panel/market-trend/MarketTrendPanel.tsx:235]: my hack to get things kinda working

File: public/app/plugins/panel/market-trend/MarketTrendPanel.tsx [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+      tweakScale={tweakScale}
+    >
+      {(config, alignedDataFrame) => {
+        return (
+          <>
+            <ZoomPlugin config={config} onZoom={onChangeTimeRange} />
+            <TooltipPlugin
+              data={alignedDataFrame}
+              config={config}
+              mode={TooltipDisplayMode.Multi}
```
Comment (ryantxu) [inline public/app/plugins/panel/market-trend/MarketTrendPanel.tsx:246]: @leeoniya note -- i forced this to always multi

File: public/app/plugins/panel/market-trend/utils.ts [hunk 3]
```diff
@@ -0,0 +1,167 @@
+import { MarketTrendMode, ColorStrategy, PriceStyle } from './types';
+import uPlot from 'uplot';
+import { colorManipulator } from '@grafana/data';
+
+const { alpha } = colorManipulator;
+
+export function drawMarkers(opts) {
+  let { mode, priceStyle, fields, colorStrategy, upColor, downColor, flatColor, volumeAlpha, flatAsUp = true } = opts;
```
Comment (ryantxu) [inline public/app/plugins/panel/market-trend/utils.ts:8]: @leeoniya -- this is the function that is the biggest blocker for types, can you take a stab at it?  The types can be defined inline and just includes types so the rest of us mortals have a chance

```suggestion
export function drawMarkers({ 
  mode: ???, 
  priceStyle: ???, 
  fields: ???, 
  colorStrategy, 
  upColor, 
  downColor, 
  flatColor, 
  volumeAlpha, 
  flatAsUp = true } ) {
```

File: public/app/plugins/panel/market-trend/MarketTrendPanel.tsx (hunk 1 above)
Comment (leeoniya) [reply public/app/plugins/panel/market-trend/MarketTrendPanel.tsx]: fixed

File: packages/grafana-ui/src/components/GraphNG/GraphNG.tsx [hunk 4]
```diff
@@ -41,12 +43,16 @@ export interface GraphNGProps extends Themeable2 {
...
+  renderers?: Renderers;
+  tweakScale?: (opts: ScaleProps) => ScaleProps;
+  tweakAxis?: (opts: AxisProps) => AxisProps;
   onLegendClick?: (event: GraphNGLegendEvent) => void;
   children?: (builder: UPlotConfigBuilder, alignedFrame: DataFrame) => React.ReactNode;
   prepConfig: (alignedFrame: DataFrame, allFrames: DataFrame[], getTimeRange: () => TimeRange) => UPlotConfigBuilder;
   propsToDiff?: Array<string | PropDiffFn>;
   preparePlotFrame?: (frames: DataFrame[], dimFields: XYFieldMatchers) => DataFrame;
   renderLegend: (config: UPlotConfigBuilder) => React.ReactElement | null;
+  options?: Record<string, any>;
```
Comment (ryantxu) [inline packages/grafana-ui/src/components/GraphNG/GraphNG.tsx:55]: Do i understand right that this is not actually used for anything -- only as trigger to `propsToDiff`?  If that is the case, lets make a big comment and maybe @dprokop has a better suggestion

Comment (leeoniya) [reply packages/grafana-ui/src/components/GraphNG/GraphNG.tsx:55]: yes, this is correct.

**Review by ryantxu — APPROVED** [review]
I think this is ready to merge, but want to continue to improve it as an alpha panel.  In particular, the setup experience needs to be a bit better (ie automatically work with fields named "open" "high", ...)

It is good that we allow anything timeseries can do, but I think this panel is also a good opportunity to explore how to setup bands more intuitively

**Comment by ryantxu** [comment]
@dprokop -- can you check this out and see if anything looks too surprising?  I don't like the 'options' field... but in general want to move all the `propsToDiff()` stuff so the panel controls everything!

**Comment by torkelo** [comment]
Can we add a devenv test dashboard as well with some test cases?

**Comment by torkelo** [comment]
ah I see https://github.com/grafana/grafana/pull/41334

**Pull request merged**